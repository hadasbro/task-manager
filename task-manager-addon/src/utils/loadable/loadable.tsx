import React, { ComponentProps, ComponentType, lazy, ReactNode, Suspense } from 'react';
import { Unpromisify } from '../../types/templates/Unpromisify';

/**
 * Opts
 */
interface Opts {
  fallback: ReactNode;
}

/**
 * lazyLoad
 *
 * @param importFunc
 * @param selectorFunc
 * @param opts
 */
export const lazyLoad = <T extends Promise<any>, U extends ComponentType<any>>(
  importFunc: () => T,
  selectorFunc?: (s: Unpromisify<T>) => U,
  opts: Opts = { fallback: null },
) => {
  let lazyFactory: () => Promise<{ default: U }> = importFunc;

  if (selectorFunc) {
    lazyFactory = () => importFunc().then(module => ({ default: selectorFunc(module) }));
  }

  const LazyComponent = lazy(lazyFactory);

  return (props: ComponentProps<U>): JSX.Element => (
    <Suspense fallback={opts.fallback!}>
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <LazyComponent {...props} />
    </Suspense>
  );
};
