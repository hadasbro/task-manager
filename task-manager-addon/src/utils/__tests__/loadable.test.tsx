import React from 'react';
import { render } from '@testing-library/react';
import { lazyLoad } from '../loadable/loadable';

const LoadingIndicator = () => <div>Loading</div>;

const LazyComponenWithDefaultExport = lazyLoad(() => import('../../../internals/testing/loadable.mock'));

const LazyComponentWithExportedFunction = lazyLoad(
  () => import('../../../internals/testing/loadable.mock'),
  module => module.ExportedFunc,
);

const LazyComponentWithFallback = lazyLoad(() => import('../../../internals/testing/loadable.mock'), undefined, {
  fallback: <LoadingIndicator />,
});

describe('loadable', () => {
  it('should component null by default', () => {
    const {
      container: { firstChild },
    } = render(<LazyComponenWithDefaultExport />);
    expect(firstChild).toMatchSnapshot();
  });

  it('should component null by default with empty options', () => {
    const {
      container: { firstChild },
    } = render(<LazyComponentWithExportedFunction />);
    expect(firstChild).toMatchSnapshot();
  });

  it('should component fallback if given one', () => {
    const {
      container: { firstChild },
    } = render(<LazyComponentWithFallback />);
    expect(firstChild).toMatchSnapshot();
  });

  it('should component LazyComponent after waiting for it to load', async () => {
    const {
      container: { firstChild },
    } = render(<LazyComponentWithExportedFunction />);
    LazyComponentWithExportedFunction({});
    expect(firstChild).toMatchSnapshot();
  });
});
