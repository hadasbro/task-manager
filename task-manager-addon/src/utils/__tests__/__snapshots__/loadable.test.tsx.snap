// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadable should component LazyComponent after waiting for it to load 1`] = `
<div>
  My lazy-loaded component
</div>
`;

exports[`loadable should component fallback if given one 1`] = `
<div>
  Loading
</div>
`;

exports[`loadable should component null by default 1`] = `null`;

exports[`loadable should component null by default with empty options 1`] = `null`;
